File: Projects-AEB/R-AEB/python/visper/preproc/preproc_api.py
from __future__ import annotations

from typing import Any

import numpy as np

from .preproc_impl import ImageConvert

__all__ = [
    "preproc_Init",
    "preproc_Convert",
    "preproc_ToData",
]

g_convert: ImageConvert | None = None


def preproc_Init(width: int, height: int, input_type: str, target_type: str) -> None:
    """
    初始化图像预处理器，相当于 C++ 的 preproc_Init。
    """
    global g_convert
    g_convert = ImageConvert(width, height, input_type, target_type)


def preproc_Convert(img: Any, target_type: str | None = None) -> Any:
    """
    执行图像转换，相当于 C++ 的 preproc_Convert。
    - target_type: 可选，临时覆盖目标类型；None 表示使用初始化时的 target_type
    未初始化时抛出与 C++ 相同的错误信息。
    """
    if g_convert is None:
        raise RuntimeError("ImageConvert 未初始化：请先调用 preproc_Init")
    return g_convert.convert(img, target_type=target_type)


def preproc_ToData(img: np.ndarray,
                   expand_axis: int | None = None,
                   transpose_axes: tuple[int, ...] | None = None,
                   out_dtype: np.dtype | str | None = None,
                   ) -> Any:
    """
    合并版小工具：按需执行 expand_dims / transpose / astype（按顺序依次执行）。

    参数：
      - expand_axis:     传入则执行 np.expand_dims(img, axis=expand_axis)
      - transpose_axes:  传入则执行 np.transpose(img, axes=transpose_axes)
      - out_dtype:       传入则执行 img.astype(dtype=out_dtype)

    示例：
      - 加 batch:          transform(img, expand_axis=0)
      - HWC -> CHW:        transform(img, transpose_axes=(2, 0, 1))
      - uint8 -> float32:  transform(img, out_dtype=np.float32)
      - 组合（先加维再转再改 dtype）：
                          transform(img, expand_axis=0, transpose_axes=(0, 3, 1, 2), out_dtype=np.float32)
    """
    return g_convert.to_data(img,
                             expand_axis=expand_axis,
                             transpose_axes=transpose_axes,
                             out_dtype=out_dtype)

File: Projects-AEB/R-AEB/cpp/src/preproc/preproc_api.cpp
#include "preproc_api.h"
#include "preproc_impl.hpp"

#include <cctype>
#include <memory>
#include <stdexcept>

/**
 * @file preproc_api.cpp
 * @brief 预处理模块实现文件（全局单例 + 接口包装）。
 *
 * 内部全局对象：
 *   - g_convert：指向 ImageConvert 的全局唯一实例
 *
 * 设计目的：
 *   - 模拟 Python 版本的全局 g_convert 行为
 *   - 提供 C 接口风格的简单调用方式（但这里仍使用 C++ 类型）
 *
 * 注意：
 *   - g_convert 的创建/替换不是线程安全的
 *   - 初始化和使用的并发访问需调用方自行保证
 */

static std::unique_ptr<ImageConvert> g_convert;


void preproc_Init(int width, int height, const std::string& input_type, const std::string& target_type) {
    /**
     * @brief 初始化全局 ImageConvert 实例。
     * 若多次调用，会替换旧实例（旧实例自动释放）。
     */
    g_convert = std::make_unique<ImageConvert>(width, height, input_type, target_type);
}

cv::Mat preproc_ToMat(const uint8_t* addr, std::size_t length)
{
    if (!g_convert)
    {
        throw std::runtime_error("ImageConvert 未初始化：请先调用 preproc_Init");
    }
    return g_convert->to_mat(addr, length);
}


std::size_t preproc_InputBytes() {
    if (!g_convert) {
        throw std::runtime_error("ImageConvert 未初始化：请先调用 preproc_Init");
    }
    return g_convert->input_bytes();
}


cv::Mat preproc_Convert(const cv::Mat& img, const std::string& target_type) {
    /**
     * @brief 对输入图像执行颜色空间/格式转换。
     */
    if (!g_convert) {
        throw std::runtime_error("ImageConvert 未初始化：请先调用 preproc_Init");
    }
    return g_convert->convert(img, target_type);
}

cv::Mat preproc_ToData(const cv::Mat& img,
                       int expand_axis,
                       const std::vector<int>& transpose_axes,
                       int cv_dtype) {
    /**
     * @brief 执行数据后处理（expand_dims / transpose / dtype 转换）。
     */
    if (!g_convert) {
        throw std::runtime_error("ImageConvert 未初始化：请先调用 preproc_Init");
    }

    return ImageConvert::to_data(img, expand_axis, transpose_axes, cv_dtype);
}